File: scripts/generate-icons.js
#!/usr/bin/env node
/**
 * Generate branded FillRight PWA icons using sharp.
 *
 * Renders the Lucide Fuel SVG icon on a purple background at multiple sizes:
 *   icon-192.png            (192×192) — Android PWA, purpose: any
 *   icon-512.png            (512×512) — Android PWA, purpose: any
 *   icon-192-maskable.png   (192×192) — artwork within W3C maskable safe-zone
 *   icon-512-maskable.png   (512×512) — artwork within W3C maskable safe-zone
 *   apple-touch-icon.png    (180×180) — iOS Add to Home Screen
 *
 * Run from the repo root:
 *   node scripts/generate-icons.js
 */

const sharp = require("sharp");
const path = require("path");
const fs = require("fs");

const PURPLE = "#7c3aed";

// Lucide Fuel icon paths (24×24 viewBox, ISC license)
const FUEL_PATHS = [
  "M14 13h2a2 2 0 0 1 2 2v2a2 2 0 0 0 4 0v-6.998a2 2 0 0 0-.59-1.42L18 5",
  "M14 21V5a2 2 0 0 0-2-2H5a2 2 0 0 0-2 2v16",
  "M2 21h13",
  "M3 9h11",
];

/**
 * Build an SVG string for the icon at 512×512.
 *
 * Non-maskable: content at ~70% of canvas (scale 15 = 360px, 76px padding).
 * Maskable: content fits within the W3C safe-zone circle (radius = 40% of
 * canvas = 204.8px → largest inscribed square ≈ 289px → scale 12 = 288px,
 * 112px padding).
 */
function makeSvg(maskable = false) {
  const scale = maskable ? 12 : 15;
  const contentSize = 24 * scale;
  const padding = (512 - contentSize) / 2;
  const pathEls = FUEL_PATHS.map((d) => `    <path d="${d}"/>`).join("\n");

  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512" width="512" height="512">
  <rect width="512" height="512" fill="${PURPLE}"/>
  <g transform="translate(${padding} ${padding}) scale(${scale})" stroke="#ffffff" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" fill="none">
${pathEls}
  </g>
</svg>`;
}

async function main() {
  const publicDir = path.join(__dirname, "..", "public");

  const targets = [
    { name: "icon-192.png", size: 192, maskable: false },
    { name: "icon-512.png", size: 512, maskable: false },
    { name: "icon-192-maskable.png", size: 192, maskable: true },
    { name: "icon-512-maskable.png", size: 512, maskable: true },
    { name: "apple-touch-icon.png", size: 180, maskable: false },
  ];

  for (const { name, size, maskable } of targets) {
    const svg = Buffer.from(makeSvg(maskable));
    const out = path.join(publicDir, name);
    await sharp(svg).resize(size, size).png().toFile(out);
    const bytes = fs.statSync(out).size;
    const variant = maskable ? "maskable" : "any    ";
    console.log(
      `  [${variant}]  ${String(size).padStart(3)}px  ${name.padEnd(30)}  ${bytes.toLocaleString()} bytes`
    );
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/generate-icons.py
#!/usr/bin/env python3
"""
Generate branded FillRight PWA icons from scratch using only Python built-ins.

Renders a fuel-pump silhouette on an orange background at multiple sizes:
  icon-192.png       (192×192) — Android PWA
  icon-512.png       (512×512) — Android PWA / splash
  apple-touch-icon.png (180×180) — iOS Add to Home Screen

Run from the repo root:
  python3 scripts/generate-icons.py
"""

import struct
import zlib
import math
import os

# ── Brand colours ────────────────────────────────────────────────────────────
ORANGE = (249, 115, 22)   # #f97316
WHITE  = (255, 255, 255)
TRANSPARENT = (0, 0, 0, 0)

# ── PNG helpers ───────────────────────────────────────────────────────────────

def _make_chunk(tag: bytes, data: bytes) -> bytes:
    crc = zlib.crc32(tag + data) & 0xFFFFFFFF
    return struct.pack(">I", len(data)) + tag + data + struct.pack(">I", crc)


def encode_png(pixels: list[list[tuple]], width: int, height: int) -> bytes:
    """Encode a 2-D list of (R,G,B) rows into a PNG bytestring."""
    raw = bytearray()
    for row in pixels:
        raw.append(0)  # filter type: None
        for r, g, b in row:
            raw += bytes([r, g, b])

    ihdr = struct.pack(">IIBBBBB", width, height, 8, 2, 0, 0, 0)
    return (
        b"\x89PNG\r\n\x1a\n"
        + _make_chunk(b"IHDR", ihdr)
        + _make_chunk(b"IDAT", zlib.compress(bytes(raw), 9))
        + _make_chunk(b"IEND", b"")
    )


# ── Geometry helpers ──────────────────────────────────────────────────────────

def lerp(a, b, t):
    return a + (b - a) * t


def blend(src: tuple, dst: tuple, alpha: float) -> tuple:
    """Alpha-composite src over dst."""
    return (
        round(lerp(dst[0], src[0], alpha)),
        round(lerp(dst[1], src[1], alpha)),
        round(lerp(dst[2], src[2], alpha)),
    )


def paint_rect(buf, bw, bh, x, y, w, h, rx, color):
    """
    Paint an axis-aligned rounded rectangle onto buf (list of lists).
    Uses supersampled anti-aliasing (4×4 sub-pixels per pixel).
    x, y, w, h, rx are in *buffer* coordinates.
    """
    SS = 4  # supersampling factor
    x2, y2 = x + w, y + h

    for py in range(max(0, int(y) - 1), min(bh, int(y2) + 2)):
        for px in range(max(0, int(x) - 1), min(bw, int(x2) + 2)):
            hits = 0
            for sy in range(SS):
                for sx in range(SS):
                    # Sample point in continuous coords
                    fx = px + (sx + 0.5) / SS
                    fy = py + (sy + 0.5) / SS

                    # Is sample inside the rounded rect?
                    if fx < x or fx > x2 or fy < y or fy > y2:
                        continue

                    # Corner proximity check
                    cx_left  = x  + rx
                    cx_right = x2 - rx
                    cy_top   = y  + rx
                    cy_bot   = y2 - rx

                    in_corner = False
                    for (qx, qy) in [(cx_left, cy_top), (cx_right, cy_top),
                                     (cx_left, cy_bot),  (cx_right, cy_bot)]:
                        if fx < (qx - rx) or fx > (qx + rx):
                            continue
                        if fy < (qy - rx) or fy > (qy + rx):
                            continue
                        # Only check if we're in the "corner box"
                        if (fx < cx_left or fx > cx_right) and (fy < cy_top or fy > cy_bot):
                            dist = math.hypot(fx - qx, fy - qy)
                            if dist > rx:
                                in_corner = True
                            break
                    if not in_corner:
                        hits += 1

            if hits > 0:
                alpha = hits / (SS * SS)
                buf[py][px] = blend(color, buf[py][px], alpha)


# ── Icon renderer ─────────────────────────────────────────────────────────────

def render(size: int) -> list[list[tuple]]:
    """
    Render the FillRight fuel-pump icon at `size`×`size`.

    All design coordinates are in a 512×512 space and scaled uniformly.
    """
    s = size / 512  # scale factor

    # Initialise canvas to solid orange
    buf = [[ORANGE] * size for _ in range(size)]

    def rect(x, y, w, h, rx, color=WHITE):
        paint_rect(buf, size, size,
                   x * s, y * s, w * s, h * s,
                   rx * s, color)

    # Pump cabinet (body)
    rect(156, 75, 200, 260, 16)

    # Display screen on cabinet face
    rect(184, 112, 144, 88, 10, ORANGE)

    # Pump base / stand
    rect(126, 318, 260, 76, 14)

    # Hose arm leaving right side of cabinet
    rect(356, 138, 65, 22, 11)

    # Vertical hose drop
    rect(399, 138, 22, 108, 11)

    # Nozzle handle (horizontal)
    rect(308, 224, 113, 22, 11)

    # Nozzle spout (vertical, at left end of handle)
    rect(300, 214, 22, 58, 11)

    return buf


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    public = os.path.join(os.path.dirname(os.path.dirname(__file__)), "public")

    targets = [
        ("icon-192.png",        192),
        ("icon-512.png",        512),
        ("apple-touch-icon.png", 180),
    ]

    for filename, size in targets:
        buf = render(size)
        png = encode_png(buf, size, size)
        out = os.path.join(public, filename)
        with open(out, "wb") as f:
            f.write(png)
        print(f"  wrote {out}  ({len(png):,} bytes)")


if __name__ == "__main__":
    main()
